### PR TITLE
Allow user specific themes

### PR DESCRIPTION
**Pull request by SourceDoctor**
users can select their Theme as they want

### CODE REVIEW

File: app/Http/Controllers/UserPreferencesController.php [hunk 1]
```diff
@@ -118,4 +129,14 @@ private function getValidLocales()
             }
         }, []);
     }
+
+    private function getValidStyles()
+    {
+        $config_data = json_decode(file_get_contents(Config::get('install_dir') . '/misc/config_definitions.json'), true)['config'];
```
Comment (Jellyfrog) [inline app/Http/Controllers/UserPreferencesController.php]: Use Config::getDefinitions()

File: resources/views/layouts/librenmsv1.blade.php [hunk 2]
```diff
@@ -40,8 +40,11 @@
     <link href="{{ asset('css/select2.min.css') }}" rel="stylesheet" type="text/css" />
     <link href="{{ asset('css/select2-bootstrap.min.css') }}" rel="stylesheet" type="text/css" />
     <link href="{{ asset('css/query-builder.default.min.css') }}" rel="stylesheet" type="text/css" />
-    <link href="{{ asset(LibreNMS\Config::get('stylesheet', 'css/styles.css')) }}?ver=20190912" rel="stylesheet" type="text/css" />
-    <link href="{{ asset('css/' . LibreNMS\Config::get('site_style', 'light') . '.css?ver=632417642') }}" rel="stylesheet" type="text/css" />
+    <link href="{{ asset(LibreNMS\Config::get('stylesheet', 'css/styles.css')) }}?ver=20190912" rel="stylesheet" type="text/css" />@if(Auth::user())
```
Comment (Jellyfrog) [inline resources/views/layouts/librenmsv1.blade.php]: Newline plz :)

Comment (SourceDoctor) [reply resources/views/layouts/librenmsv1.blade.php]: ups, fixed ;-)

File: app/Http/Controllers/UserPreferencesController.php (hunk 1 above)
Comment (SourceDoctor) [reply app/Http/Controllers/UserPreferencesController.php]: done :-)

**Comment by SourceDoctor** [comment]
@Jellyfrog 
everything changed ;-)

File: app/Http/Controllers/UserPreferencesController.php [hunk 3]
```diff
@@ -57,8 +57,11 @@ public function index(Request $request)
             'dashboards' => Dashboard::allAvailable($user)->with('user')->get(),
             'default_dashboard' => UserPref::getPref($user, 'dashboard'),
             'note_to_device' => UserPref::getPref($user, 'add_schedule_note_to_device'),
-            'locale' => UserPref::getPref($user, 'locale') ?: 'en',
+            'locale' => UserPref::getPref($user, 'locale') ?: Config::get('locale'),
```
Comment (murrant) [inline app/Http/Controllers/UserPreferencesController.php]: Config::get('local') is not the right setting.  It should be `\config('app.locale', 'en');`

File: app/Http/Controllers/UserPreferencesController.php [hunk 4]
```diff
@@ -118,4 +129,13 @@ private function getValidLocales()
             }
         }, []);
     }
+
+    private function getValidStyles()
+    {
+        $site_styles = Config::getDefinitions()['site_style']['options'];
```
Comment (murrant) [inline app/Http/Controllers/UserPreferencesController.php]: Best way is:
```php
$definitions = new DynamicConfig();
return $defintions->get('site_style')->getOptions();
```

Comment (murrant) [reply app/Http/Controllers/UserPreferencesController.php]: Why are you sorting them? I'd suggest re-ordering them in the definition instead.

File: app/Http/ViewComposers/MenuComposer.php [hunk 5]
```diff
@@ -54,7 +55,8 @@ public function compose(View $view)
         /** @var User $user */
         $user = Auth::user();
 
-        $vars['navbar'] = in_array(Config::get('site_style'), ['mono', 'dark']) ? 'navbar-inverse' : '';
+        //Todo: should be handled via CSS Themes
+        $vars['navbar'] = in_array(UserPref::getPref($user, 'site_style'), ['mono', 'dark']) ? 'navbar-inverse' : '';
```
Comment (murrant) [inline app/Http/ViewComposers/MenuComposer.php]: This does not fall back to the global setting.

**Comment by murrant** [comment]
Couple of items:
1. every time you get the style it should be user -> global
2. UserPref::getPref() is an SQL query everytime you call it, so try to minimize those queries.

File: app/Http/Controllers/UserPreferencesController.php (hunk 3 above)
Comment (SourceDoctor) [reply app/Http/Controllers/UserPreferencesController.php]: changed :-)

File: app/Http/Controllers/UserPreferencesController.php (hunk 4 above)
Comment (SourceDoctor) [reply app/Http/Controllers/UserPreferencesController.php]: changed :-)

File: app/Http/ViewComposers/MenuComposer.php (hunk 5 above)
Comment (SourceDoctor) [reply app/Http/ViewComposers/MenuComposer.php]: missing fallback implemented

**Comment by SourceDoctor** [comment]
@murrant 
things changed :-)

**Comment by murrant** [comment]
The user preference handling needed a little polishing on top of this :)

**Comment by SourceDoctor** [comment]
@murrant 
in user preferences 
default language is always the configured language ;-)

**Comment by murrant** [comment]
Not in my install, it is working fine.
valid, I'll create a fix.
There are no language settings in global settings

**Comment by SourceDoctor** [comment]
@murrant 
sorry i meant css stylenot language :o)

**Comment by murrant** [comment]
Are you still having issues

**Comment by SourceDoctor** [comment]
yes, every CSS Style i configure in User Management is preselected in Global Settings

**Comment by murrant** [comment]
that is already fixed... #10820

**Comment by SourceDoctor** [comment]
@murrant 
i see this fix was merged yesterday. i tested it and yo're right. It's fixed now. :-)

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-58-release-changelog-november-2019/10224/1

**Pull request merged**